### ISSUE TITLE
Current unability of LuaLaTeX from TeXLive 2016 pretest to handle ttc fonts (on mac os x)

### ISSUE THREAD

**Issue by jfbu** [lualatex/luaotfload#328]
Hi,

not sure this is the right place to report. But I reproduce here a message I sent to
 tex-live mailing  list which is possibly not the best place, as pointed out by
its administrators.

This test file

```latex
\documentclass{article}

\usepackage{fontspec}

\setmainfont[BoldFont={* Black}]{Hoefler Text}

%\setmainfont[BoldFont={* Condensed Medium}]{Futura}

%\setmainfont[BoldFont={* Bold}]{Gill Sans}

\begin{document}

Hello

\bfseries

World

\end{document}
```

fails to compile successfully when any of the three lines is used.

All three fonts are system fonts on a mac os x system (Mavericks, 10.9.5).
These fonts came with the mac os.

They all share the fact to be fonts with extension `.ttc`. The main font 
is found but not the variants.

There is no problem with TeXLive 2015. The problem shows with current
TeXLive 2016 which is in pretest. There is no problem with XeLaTeX,
be it in TL2015 or pretest 2016.

I can provide additional information (logs etc...). I have not tested
beyond these three fonts, but I could test a few more, I am ready to bet
all `.ttc` fonts will exhibit the phenomenon.

I have noticed in passing a fontspec issue which I believe has been
already fixed in the dev repo today by Joseph. This happens
with Scale=MatchLowercase.

It is quite possible this issue I am creating here is silly because TL pretest
in a state of flux regarding LuaTeX, and  luaotfload has not been
yet uploaded there, and the upstream has already solved all those
problems, and the only thing is that I am with a luaotfload for the
LuaTeX of  TL2015 and I am testing too early with the LuaTeX of TL2016.

**Comment by phi-gamma** [comment]
···<date: 2016-04-11, Monday>···<from: jfbu>···

> not sure this is the right place to report.

No, it’s definitely alright =) We’ll sort it out eventually.

Regarding the report:

> All three fonts are system fonts on a mac os x system (Mavericks, 10.9.5).
> These fonts came with the mac os.

Those can’t be shared, so the report isn’t reproducible. If you
can reproduce it with a free font hat would help. An example that
fails out of the box with some font shipped with tex live would
be valuable for regression testing later on.

> They all share the fact to be fonts with extension `.ttc`. The main font 
> is found but not the variants.

There is one possible … ahem “disagreement” with the current code
for subfont indexing with Luatex’s assumptions. It’s already gone
in my development tree and we’ve consulted with upstream about
the right indexing method, so there’s a fair chance that this bug
will just vanish in the next version.

> I have noticed in passing a fontspec issue which I believe has been
> already fixed in the dev repo today by Joseph. This happens
> with Scale=MatchLowercase.

Ah, that might possible too.

**Comment by jfbu** [comment]
There are no fonts distributed with TeX Live in "TrueType Collection" format, as far as I know.

As I am not working with fonts, it wasn't easy for me to identify trusted places on the net from where I could download TTC fonts. I went to http://arkandis.tuxfamily.org/adffonts.html. But could find only one available download coming with TTC font files: **IRIANIS**. It is a free font with a GNU license. There is  a zip file.

    http://arkandis.tuxfamily.org/adffonts.html

Turns out however that the archive contains four TTC files
```
IrianisADFStd-Bold.ttc
IrianisADFStd-Regular.ttc
IrianisADFStd-BoldItalic.ttc
IrianisADFStd-Italic.ttc
```
and each  TTC file in fact encapsulates two fonts, the alternative having an extra `Style` in its name. This was revealed from installing the font on my Mac, and checking in the application FontBook.

Thus to reproduce the issue reported here I used that source file:

```latex
\documentclass{article}

\usepackage{fontspec}
\setmainfont[BoldFont={IrianisADFStyleStd-Regular}, 
             UprightFont={IrianisADFStd-Regular}]{Irianis ADF Std}

\begin{document}

Hello

\bfseries

World

\end{document}
```

It mimicks the earlier situation with the Bold variant being inside the TTC encapsulation.

### As expected XeLaTeX has no problem but LuaLaTeX does not find the "Style" variant.

There is a regression here because the TeXLive 2015 has no problem. The above compiles fine with LuaLaTeX from TL2015, the compilation console output ending with:

```
</Users/xxxxxx/Library/Fonts/IrianisADFStd-Regular.ttc(IrianisADFStyleStd-Regular:1)></Users/xxxxxx/Library/Fonts/IrianisADFStd-Regular.ttc(IrianisADFStd-Regular:0)>Output written on testirianisluatex.pdf (1 page, 9660 bytes).
```

and starting with 
```
This is LuaTeX, Version beta-0.80.0 (TeX Live 2015) (rev 5238) 
 restricted \write18 enabled.
```

and contains
```
luaotfload | db : Reload initiated (formats: otf,ttf,ttc,dfont); reason: "Font IrianisADFStd-Regular not found.".(compiling
```
showing that the font cache was successfully redone on the fly (following my recent install of the font in `$HOME/Library/Fonts`.) In the TL2016 case I naturally did a manual 

    luaotfload-tool -u -vvv

to make for the best conditions. But to no avail.

By the way, 

```latex
\documentclass{article}

\usepackage{fontspec}

\setmainfont{Irianis ADF Std}
\setsansfont{Irianis ADF Style Std}
\begin{document}

Hello

\bfseries

World

\sffamily

Hello

\bfseries

World

\end{document}
```

has no issue with XeLaTeX, but the second line fails with LuaLaTeX from TL2016.

However both are OK on TeXLive 2015. Apart from LuaLaTeX having a kerning problem with W of World in the bold face.

Should I provide the logs, please tell so.

**Comment by phi-gamma** [comment]
···<date: 2016-04-12, Tuesday>···<from: jfbu>···

> There are no fonts distributed with TeX Live in "TrueType Collection" format, as far as I know.

Fwiw you can test Asana Math.

> ```latex
> \documentclass{article}
> 
> \usepackage{fontspec}
> \setmainfont[BoldFont={IrianisADFStyleStd-Regular}, 
>              UprightFont={IrianisADFStd-Regular}]{Irianis ADF Std}
> 
> \begin{document}
> 
> Hello
> 
> \bfseries
> 
> World
> 
> \end{document}
> ```

With 2.7, this example looks like this: http://i.imgur.com/9tSiAYo.png

> ### As expected XeLaTeX has no problem but LuaLaTeX does not find the "Style" variant.

http://commonmark.org/help/

> There is a regression here because the TeXLive 2015 has no problem.

There has been no Luaotfload release for anything later than
texlive 2015 yet.

> ```latex
> \documentclass{article}
> 
> \usepackage{fontspec}
> 
> \setmainfont{Irianis ADF Std}
> \setsansfont{Irianis ADF Style Std}
> \begin{document}
> 
> Hello
> 
> \bfseries
> 
> World
> 
> \sffamily
> 
> Hello
> 
> \bfseries
> 
> World
> 
> \end{document}
> ```


> has no issue with XeLaTeX, but the second line fails with LuaLaTeX from TL2016.

The second line is empty. But again, the example outputs this
with 2.7: http://i.imgur.com/rdrsJwK.jpg

**Comment by jfbu** [comment]
> Fwiw you can test Asana Math.

ok, but I was looking for a text font.

> With 2.7, this example looks like this: http://i.imgur.com/9tSiAYo.png

problem fixed then.

> http://commonmark.org/help/

I wanted a slightly bigger font size and had no time. BTW I didn't know github was using commonmark.

> There has been no Luaotfload release for anything later than
> texlive 2015 yet.

as said on top of thread, it is hard for end-users to know where to report such problems. As I hit a problem under LuaLaTeX with fonts not being found, I picked up luaotfload as the place to report.

> The second line is empty.

No the second line was not empty. I was not clear but I meant the `\setsansfont{Irianis ADF Style Std}` line.

>  But again, the example outputs this
> with 2.7: http://i.imgur.com/rdrsJwK.jpg
> 

Looks like it is fixed then. Note: a kerning problem with the regular font between W and o exists in TL2015 version, but not in your posted image.

I am glad something has been fixed, I don't know for sure what "upstream" means here. I hope the fix will make it to TL2016, else you might expect avalanche of reports by end users.

**Comment by phi-gamma** [comment]
···<date: 2016-04-19, Tuesday>···<from: jfbu>···
> I don't know for sure what "upstream" means here.

There’s a chapter “The Fontloader” in the manual, please refer to
that.

> I hope the fix will make it to TL2016, else you might expect
> avalanche of reports by end users.

That’s going to happen either way but I certainly wouldn’t mind
the extra feedback.

**Comment by jfbu** [comment]
see https://github.com/lualatex/luaotfload/issues/327#issuecomment-213286413

**Comment by jfbu** [comment]
To confirm all is fine now (TL 2016 pretest, April 24, `luaotfload 2.7`) after applying fix about `names-version = 2.7` mentioned at #337. Both for the original mwe with `Gill Sans` and the one with `Irianis ADF`, and I now presume for all `.ttc` encapsulated fonts.
